### DISCUSSION TITLE
Proper shutdown procedure

### DISCUSSION THREAD

**Discussion by kirirururu** [resiprocate/resiprocate#249] [Q&A] [answered]
Hi. I have a strange problem with DUM. My steps are as follows:
1. Create stack, DUM, etc.
2. Send REGISTER request
3. Wait for ClientRegistrationHandler::onSuccess() to be called
4. Call DialogUsageManager::shutdown()
5. Wait for DumShutdownHandler::onDumCanBeDeleted() to be called.

And onDumCanBeDeleted() never gets called! In log I see these messages:
```
[INFO ] (resip.dum) shutdown: dialogSets=1
[DEBUG] (resip.dum) Shutdown waiting for all usages to be deleted (2)
[DEBUG] (resip.dum) 2 -> ClientRegistration sip:101@sip.local
[DEBUG] (resip.dum) 1 -> AppDialogSet 1
[INFO ] (resip.transaction) TransactionUserMessage::RequestShutdown TU: DialogUsageManager size=0
```

Do I need to do some explicit actions to properly finish/delete all created dialogs?

**Comment by sgodin** [comment, marked as answer]
Yes, you need to remove all your dialog usages before a graceful shutdown
can occur.  In your case you need to remove your registration.

See dum/test/basicClient for an example.  In particular to code
in BasicClientUserAgent::process(int timeoutMs) which ends the usages after
a shutdown flag has been triggered.

https://github.com/resiprocate/resiprocate/wiki/DUM-Application-Shutdown

Best Regards,
Scott

**Comment by kirirururu** [comment]
Thank you. I will take a look at this example.

**Comment by kirirururu** [comment]
But what can I do with pending out-of-dialog requests?
I create OPTIONS request periodically to monitor server connection. But if the server is offline, each OPTIONS request blocks DUM shutdown for up to 32 seconds (internal timeout). How can I cancel out-of-dialog request before shutdown?

**Comment by sgodin** [comment]
There is no way in SIP to cancel a non-Invite request.  So for a
truly graceful shutdown (ie: ensuring all your transactions are fully
completed and re-transmissions are occurring according to protocol you will
need to wait up to 32 seconds.  If you are not concerned about a graceful
shutdown you can always shutdown forcefully.   I've seen some people use a
hybrid approach, ie:  start a graceful shutdown, but time it out after 5-10
seconds, then shutdown forcefully.

Scott
